packages/providers/src/resource.ts

```diff
@@ -3,28 +3,39 @@ import type { BytesLike } from '@ethersproject/bytes';
 import type { Coin } from './coin';
 import type { MessageCoin } from './message';
 
-export type RawCoin = {
+export interface RawCoin {
   utxoId: string;
   owner: string;
   amount: string;
   assetId: string;
   maturity: string;
   blockCreated: string;
   txCreatedIdx: string;
-};
+}
 
-export type RawMessage = {
+export interface RawMessage {
   amount: string;
   sender: string;
   assetId: string;
   recipient: string;
   data: string;
   nonce: string;
   daHeight: string;
-};
+}
+
+export interface CoinPredicate extends Coin {
+  getPredicateContent: () => { predicate: Uint8Array; predicateData: Uint8Array };
+}
+
+export interface MessageCoinPredicate extends MessageCoin {
+  getPredicateContent: () => { predicate: Uint8Array; predicateData: Uint8Array };
+}
```

Comment:
Why do we need this method? 

We can simply make `Coin` and `MessagePredicate` extend `Coin` and `MessageCoin` and define both `predicate` and `predicateData` props.

Reply:
If we assign `predicate` and `predicateData` to a predicate's resources the moment we fetch it in `getResourcesToSpend`, updating that predicate's data via `predicate.setData` before adding those resources to a transaction request  won't affect those resources' `predicateData`. Returning those properties via a callback would allow that.

```ts
const predicate = new Predicate(...);
const resources = await predicate.getResourcesToSpend(...); 

// someData would not be reflected in the resources above 
// if we set predicateData right away upon fetching,
// but with the getPredicateContent callback, 
// you can change the data as much as you like before adding it to transactionRequest,
// because it's behind a callback which holds a reference to the predicate's properties.
predicate.setData(someData);

transactionRequest.addResources(resources);
```

This idea is illustrated in the two tests added in `packages/fuel-gauge/src/predicate-resources-with-custom-transaction.test.ts`.